### ISSUE TITLE
Query Formatter for Mongo seem to be not parsable into js object

### ISSUE THREAD

**Issue by mylawacad** [react-querybuilder/react-querybuilder#256]
Hi

At the moment queries formatted with `const testMongoQueryStr = formatQuery(item, 'mongodb')` can not be further parsed with JSON.parse() i.e `const mongoQueryParams = JSON.parse(testMongoQueryStr )`  and  so are throwing an error if you would try to use it as a part of mongoDb query.

As an example please take a look at this test in the repo. As far as I understand this needs to be an object not a string and also this string needs to be valid for JSON.parse() : https://github.com/react-querybuilder/react-querybuilder/blob/e126531f2e49f90e54319e1a9cfefd35a950be96/packages/react-querybuilder/src/utils/__tests__/formatQuery.test.ts#L290

Additionally any numeric fields are formatted as strings like so using the `formatQuery(item, 'mongodb')`:  `"{$and:[{testNumInput:{$lte:\"67\"}}]}"` instead of: `"{$and:[{"testNumInput":{$lte:67}}]}"`   and $lte will not work as expected because strings will not be properly compared against numbers.

Here is an example of properly formatted mongodb query:

`const cursor = db.collection('inventory').find({
  status: 'A',
  qty: { $lt: 30 }
});`

at: https://docs.mongodb.com/manual/tutorial/query-documents/#specify-and-conditions

I would expect that formatQuery should return the following in that case: `'{"status":"A","qty":{"$lt":30}}'`  and instead it returns this: `'{status:"A",qty:{$lt:"30"}}'` which can not be parsed into js object

**Comment by jakeboone02** [comment]
Thanks for the report, @mylawacad. The `"mongodb"` format was a contributed feature (#214), and since I didn't have much experience with MongoDB I took it on faith that the export format was valid without testing against a real implementation. I think it would be good to revisit the feature and include any changes as part of the upcoming v4.0.

First we need to decide if the output should be a valid `JSON.parse`-able string, or if the output should be a proper JSON object. I'd probably prefer the object, all else equal, but I'll consider your input. In either case we'll need to convert the `/pattern/` syntax for `"contains"`, `"beginsWith"`, etc., to use the [`{"$regex": "pattern"}` syntax](https://docs.mongodb.com/manual/reference/operator/query/regex/) instead, and if we leave it as a string then we'd need to put double quotes around the identifiers. Shouldn't be too much work either way.

The numeric value issue is a little more complicated for a couple of reasons. By default, the query builder does not produce `number`-type values even when the field specifies `inputType: "number"` (which produces an `<input type="number" />` value editor) because HTML input elements always report `value` as a string. Also, this library does not have an official way to associate a datatype with a field. Therefore, `formatQuery` can't easily determine (using known properties) if a given value should be represented as a string or a number in the output. This isn't an issue with most SQL databases because they will cast strings as numbers when appropriate (e.g. if `field` is a `NUMBER` column in table `t`, then `SELECT * FROM t WHERE field = '1'` will be treated the same as `SELECT * FROM t WHERE field = 1`). But I understand it's different with NoSQL databases like MongoDB.

We have a couple of options to address this issue:

1. Have `formatQuery` produce raw numbers when a rule's `value` is a number. This would fix the issue in `formatQuery`, but the default `ValueEditor` would still produce strings. The user would need to use a custom `ValueEditor` component just to take advantage of the default `formatQuery` behavior.
2. Option 1, but also produce a numeric `value` in the default `ValueEditor` when `inputType` is `"number"`. This would be a breaking change (we're about to do a major version bump anyway, so NBD), and I'd be a little worried about unforeseen edge cases.
3. Leave the `ValueEditor` alone, and make use of the `valueProcessor` option in the `"mongodb"` code path within `formatQuery`. Right now it only uses that option in the SQL-based code paths. If we used it for `"mongodb"`, then the user could decide what was represented as a number based on the `field` and `operator` just like they can today for the SQL formats. This would not necessarily preclude Option 1, but would make it technically unnecessary.

Option 3 is the easiest to implement and would avoid any breaking changes. However, I can certainly see the benefit of the library treating numbers as numbers by default. It's probably what most users would expect, anyway.

Thoughts?

**Comment by mylawacad** [comment]
Hi @jakeboone02 

Thank you for taking a look and for the explanation.  I just made a PR to at least make the resulting string to be parsable by JSON.parse() i.e be a valid JSON. Agree with you that it should be more nice if `formatQuery()` would return an object rather then a string for the mongodb case.

Regarding the numeric type field - yes, I saw yesterday in the code that there is no number type and had a thought that things may not be so simple in that case.  I'm not yet that much familiar with the project to be able to pick some option among provided.  In any case I'm currently working over implementing react-querybuilder into a project that uses mongodb so might come up with a proposal later.  I was thinking to start from looking into "Option 3" and see if that will suit.

Just thinking out loud - Option1 might be covering most of the cases, but as you correctly mentioned mongodb will fail to compare numeric against  string, so I'm not totally sure if there are some edge cases where that will not work as expected.
